libs/s25main/ingameWindows/iwConnecting.cpp

```diff
@@ -0,0 +1,101 @@
+// Copyright (C) 2005 - 2022 Settlers Freaks (sf-team at siedler25.org)
+//
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#include "iwConnecting.h"
+#include "Loader.h"
+#include "WindowManager.h"
+#include "controls/ctrlPercent.h"
+#include "controls/ctrlText.h"
+#include "desktops/dskGameLobby.h"
+#include "iwMsgbox.h"
+#include "network/GameClient.h"
+#include "gameData/const_gui_ids.h"
+#include "s25util/colors.h"
+#include <cmath>
+
+namespace {
+enum
+{
+    ID_txtStatus,
+    ID_prgLoad,
+    ID_btAbort,
+};
+}
+
+iwConnecting::iwConnecting(ServerType serverType, std::unique_ptr<ILobbyClient> lobbyClient)
+    : IngameWindow(CGI_CONNECTING, IngameWindow::posCenter, Extent(300, 120), _("Connecting"),
+                   LOADER.GetImageN("resource", 41), true, false),
+      serverType_(serverType), lobbyClient_(std::move(lobbyClient))
+{
+    WINDOWMANAGER.SetCursor(Cursor::Moon);
+
+    AddText(ID_txtStatus, DrawPoint(150, 30), _("Please wait..."), COLOR_YELLOW, FontStyle::CENTER, NormalFont);
+    AddPercent(ID_prgLoad, DrawPoint(20, 50), Extent(260, 26), TextureColor::Grey, COLOR_YELLOW, SmallFont,
```

Comment:
using downloadProgress via the pointer is against SOLID principles (since ctrlPercent does not have control over it) but I think this is okay for this case.